### PR TITLE
fix: Add unpaid interest and delegate/treasuryFee to payment amounts (SC-5084)

### PR DESCRIPTION
**Pull request by lucas-manuel**
# Description

Moves extraFee calculations to use payments remaining post refinance as before refinance it is almost always 1.

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog

### CODE REVIEW

File: contracts/interfaces/IMapleLoan.sol [hunk 1]
```diff
@@ -130,6 +130,11 @@ interface IMapleLoan is IMapleProxied, IMapleLoanEvents {
      */
     function refinanceCommitment() external view returns (bytes32 refinanceCommitment_);
 
+    /**
+     *  @dev The pro rata principal to be paid before a refinance was accepted.
+     */
```
Comment (JGcarv) [inline contracts/interfaces/IMapleLoan.sol:135]: This description is wrong

**Review by JGcarv — APPROVED** [review]

File: contracts/interfaces/IMapleLoan.sol [hunk 2]
```diff
@@ -306,6 +311,13 @@ interface IMapleLoan is IMapleProxied, IMapleLoanEvents {
      */
     function getNextPaymentBreakdown() external view returns (uint256 principal_, uint256 interest_, uint256 delegateFee_, uint256 treasuryFee_);
 
+    /**
+     *  @dev    Get the breakdown of the principal and interest that will be charged according to loan terms before refinance.
+     *  @param  timestamp_       The timestamp when the new terms will be accepted.
+     *  @return proRataInterest_ The interest portion to be added in the next payment.
+     */
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply contracts/interfaces/IMapleLoan.sol:318]: Fixed

File: contracts/MapleLoanInternals.sol [hunk 3]
```diff
@@ -228,17 +234,45 @@ abstract contract MapleLoanInternals is MapleProxiedInternals {
 
         require(block.timestamp <= deadline_, "MLI:ANT:EXPIRED_COMMITMENT");
 
+        uint256 preRefinancePaymentInterval = _paymentInterval;
+
+        require(block.timestamp + preRefinancePaymentInterval >= _nextPaymentDueDate, "MLI:ANT:SEVERAL_PAYMENTS_DUE");
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply contracts/MapleLoanInternals.sol:239]: Fixed

File: contracts/test/MapleLoanInternals.t.sol [hunk 4]
```diff
@@ -2461,10 +2561,61 @@ contract MapleLoanInternals_AcceptNewTermsTests is TestUtils {
...
+    ) external {
+        delegateFee_       = constrictToRange(delegateFee_,       1000, 100_000);
+        treasuryFee_       = constrictToRange(treasuryFee_,       1000, 100_000);
+        globalInvestorFee_ = constrictToRange(globalInvestorFee_, 100,  1000);
+        globalTreasuryFee_ = constrictToRange(globalTreasuryFee_, 100,  1000);
+
+        _loan.setDelegateFee(delegateFee_);
+        _loan.setTreasuryFee(treasuryFee_);
+
+        _globals.setInvestorFee(globalInvestorFee_);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply contracts/test/MapleLoanInternals.t.sol:2606]: After acceptNewTerms, I have added a new test to capture this

File: contracts/MapleLoanInternals.sol [hunk 5]
```diff
@@ -228,17 +234,45 @@ abstract contract MapleLoanInternals is MapleProxiedInternals {
...
             require(success, "MLI:ANT:FAILED");
         }
 
-        _setEstablishmentFees(_principal, _paymentInterval);
+        // Track any uncaptured establishment fees and spread over the course of new loan.
+        uint256 extraDelegateFee = (_delegateFee * timeSinceLastPayment) / (preRefinancePaymentInterval * _paymentsRemaining);
+        uint256 extraTreasuryFee = (_treasuryFee * timeSinceLastPayment) / (preRefinancePaymentInterval * _paymentsRemaining);
+
+        // Increment the due date to be one full payment interval from now, to restart the payment schedule with new terms.
+        // NOTE: `_paymentInterval` here is possible newly set via the above delegate calls, so cache it.
```
Comment (edag94) [inline contracts/MapleLoanInternals.sol:270]: possibly

File: contracts/MapleLoanInternals.sol [hunk 6]
```diff
@@ -228,17 +234,45 @@ abstract contract MapleLoanInternals is MapleProxiedInternals {
...
+        _nextPaymentDueDate = block.timestamp + _paymentInterval;
+
+        // Update establishment fees.
+        // NOTE: `_principal` here is possible newly increased/decreased via the above delegate calls.
```
Comment (edag94) [inline contracts/MapleLoanInternals.sol:274]: typo: possible -> possibly

File: contracts/MapleLoan.sol [hunk 7]
```diff
@@ -257,6 +257,12 @@ contract MapleLoan is IMapleLoan, MapleLoanInternals {
         ( principal_, interest_, delegateFee_, treasuryFee_ ) = _getNextPaymentBreakdown();
     }
 
+    function getRefinanceInterest(uint256 timestamp_) external view override returns (uint256 proRataInterest) {
+        uint256 timeSinceLastPaymentDueDate = timestamp_ - (_nextPaymentDueDate - _paymentInterval);
+
+        ( , proRataInterest ) = _getInstallment(_principal, _endingPrincipal, _interestRate, timeSinceLastPaymentDueDate, _paymentsRemaining);
+    }
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply contracts/MapleLoan.sol:264]: Lets add this as part of the second PR as I think it will be related

**Review by edag94 — APPROVED** [review]

File: contracts/MapleLoanInternals.sol [hunk 8]
```diff
@@ -522,12 +521,68 @@ abstract contract MapleLoanInternals is MapleProxiedInternals {
...
+        uint256 interestRate_,
+        uint256 nextPaymentDueDate_,
+        uint256 lateFeeRate_,
+        uint256 lateInterestPremium_
+    )
+        internal pure returns (uint256 lateInterest_)
+    {
+        if (currentTime_ <= nextPaymentDueDate_) return 0;
+
+        uint256 daysLate = (((currentTime_ - nextPaymentDueDate_ - 1) / 1 days) + 1) * 1 days;
```
Comment (edag94) [inline contracts/MapleLoanInternals.sol:582]: Variable name is misleading, as its in seconds not number of days. Also clarify in a comment that all this code does is round up the time late to the nearest day.

**Review by edag94 — APPROVED** [review]

**Review by deluca-mike — APPROVED** [review]

**Pull request merged**